include/gzip/compress.hpp

```diff
@@ -41,68 +55,29 @@ class Compressor
-
-        constexpr int mem_level = 8;
-        // The memory requirements for deflate are (in bytes):
-        // (1 << (window_bits+2)) +  (1 << (mem_level+9))
-        // with a default value of 8 for mem_level and our window_bits of 15
-        // this is 128Kb
-
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wold-style-cast"
-        if (deflateInit2(&deflate_s, level_, Z_DEFLATED, window_bits, mem_level, Z_DEFAULT_STRATEGY) != Z_OK)
+        std::size_t max_compressed_size = libdeflate_gzip_compress_bound(compressor_, size);
+        // TODO: sanity check this before allocating
+        if (max_compressed_size > output.size())
         {
-            throw std::runtime_error("deflate init failed");
+            output.resize(max_compressed_size);
         }
-#pragma GCC diagnostic pop
 
-        deflate_s.next_in = reinterpret_cast<z_const Bytef*>(data);
-        deflate_s.avail_in = static_cast<unsigned int>(size);
-
-        std::size_t size_compressed = 0;
-        do
+        std::size_t actual_compressed_size = libdeflate_gzip_compress(compressor_,
+                                                                      data,
+                                                                      size,
+                                                                      reinterpret_cast<void*>(&output[0]),
+                                                                      max_compressed_size);
+        if (actual_compressed_size == 0)
         {
-            size_t increase = size / 2 + 1024;
-            if (output.size() < (size_compressed + increase))
-            {
-                output.resize(size_compressed + increase);
-            }
-            // There is no way we see that "increase" would not fit in an unsigned int,
-            // hence we use static cast here to avoid -Wshorten-64-to-32 error
-            deflate_s.avail_out = static_cast<unsigned int>(increase);
-            deflate_s.next_out = reinterpret_cast<Bytef*>((&output[0] + size_compressed));
-            // From http://www.zlib.net/zlib_how.html
-            // "deflate() has a return value that can indicate errors, yet we do not check it here.
-            // Why not? Well, it turns out that deflate() can do no wrong here."
-            // Basically only possible error is from deflateInit not working properly
-            deflate(&deflate_s, Z_FINISH);
-            size_compressed += (increase - deflate_s.avail_out);
-        } while (deflate_s.avail_out == 0);
-
-        deflateEnd(&deflate_s);
-        output.resize(size_compressed);
+            throw std::runtime_error("actual_compressed_size 0");
+        }
+        output.resize(actual_compressed_size);
     }
 };
 
 inline std::string compress(const char* data,
                             std::size_t size,
-                            int level = Z_DEFAULT_COMPRESSION)
+                            int level = 6)
```

Comment:
made this change to avoid needing to depend on zlib.h for `Z_DEFAULT_COMPRESSION`. My understanding is that inside zlib.h `Z_DEFAULT_COMPRESSION` is actually set to something like `-1` and then internally zlib flips that to mean `6` when it is actually used. libdeflate only understands `6` (or 1-N).